datarobot_batch_scoring/api_response_handlers/pred_api_v10.py

```diff
@@ -1,76 +1,207 @@
+    row_generator: iterator
+    """
+    single_row = result_sorted[0]
+    fields = sorted(
+        record['label']
+        for record in single_row['predictionValues']
+    )
     if pred_name is not None:
-        out_fields = ['row_id', pred_name]
+        headers = [pred_name]
         # batch scoring returns only positive class for binary tasks if
         # user specified pred_name option, so we eliminate negative
         # result
         if len(fields) == 2 and single_row['prediction'] in fields:
             fields = [fields[-1]]
     else:
-        out_fields = ['row_id'] + fields
+        headers = fields
+
+    rows_generator = (
+        [
+            pred_vals['value']
+            for pred_vals in
+            sorted(
+                record['predictionValues'],
+                key=operator.itemgetter('label')
+            )
+            if pred_vals['label'] in fields
+        ] for record in result_sorted
+    )
+
+    return headers, rows_generator
+
+
+def keep_original_fields(batch, keep_cols, fast_mode, input_delimiter):
+    """ Generate fields from original data
+
+    Parameters
+    ----------
+    batch
+        batch of original data
+    keep_cols: list[str]
+        columns we should keep
+    fast_mode: bool
+        faster CSV processor. if True then row in the batch is raw string
+    input_delimiter: str, optional
+        csv delimiter. should be passed if fast_mode is True
+
+    Returns
+    -------
+    header: list[str]
+    row_generator: iterator
+    """
+
+    feature_indices = {col: i for i, col in
+                       enumerate(batch.fieldnames)}
+    indices = [feature_indices[col] for col in keep_cols]
+
+    def rows_generator():
+        for original_row in batch.data:
+            if fast_mode:
+                # row is a full line, we need to cut it into fields
+                original_row = original_row.rstrip().split(input_delimiter)
+                if len(original_row) != len(batch.fieldnames):
+                    raise UnexpectedKeptColumnCount()
 
-    pred = [[record['rowId'] + batch.id] + [
-        pred_vals['value'] for pred_vals in
-        sorted(record['predictionValues'],
-               key=operator.itemgetter('label'))
-        if pred_vals['label'] in fields]
-            for record in sorted(result,
-                                 key=operator.itemgetter('rowId'))]
+            yield [original_row[i] for i in indices]
+
+    return keep_cols, rows_generator()
+
+
+def prediction_explanation_fields(
+    result_sorted,
+    prediction_explanations_key
+):
+    """ Generate prediction explanations fields
+
+    Parameters
+    ----------
+    result_sorted : list[dict]
+        list of results sorted by rowId
+    prediction_explanations_key : str
+        key of results by which explanation fields can be found
+
+    Returns
+    -------
+    header: list[str]
+    row_generator: iterator
+    """
+    headers = []
+    single_row = result_sorted[0]
```

Comment:
Will there always be at least one item in `result_sorted`? Otherwise this could raise an `IndexError`.

Reply:
Yes, there's a check that forbids sending empty dataset for scoring much earlier
https://github.com/datarobot/batch-scoring/blob/master/datarobot_batch_scoring/reader.py#L255